formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtobufDecoding.kt

```diff
@@ -225,16 +241,33 @@ internal open class ProtobufDecoder(
             if (protoId == -1) { // EOF
                 return elementMarker.nextUnmarkedIndex()
             }
-            val index = getIndexByTag(protoId)
+            val index = getIndexByNum(protoId)
             if (index == -1) { // not found
                 reader.skipElement()
             } else {
+                if (descriptor.extractParameters(index).isOneOf) {
+                    index2IdMap?.put(index, protoId)
+                }
                 elementMarker.mark(index)
                 return index
             }
         }
     }
 
+    private fun <T> decodeOneOfElement(deserializer: DeserializationStrategy<T>): T {
+        val tag = currentTagOrDefault
+        if (tag != MISSING_TAG && tag.isOneOf && deserializer is SealedClassSerializer) {
```

Comment:
It is fairly straightforward to make this also work with non-sealed types. You just need to extract the information out of the deserializer module.

Reply:
Done